File: src-tauri/src/external_recipe/sallys_welt.rs
//! This module implements [`ExternalRecipeGetterTrait`] for the domain `sallys-blog.de`.

use std::sync::{Arc, OnceLock};

use anyhow::Result;
use async_trait::async_trait;
use regex::Regex;
use url::Url;

use crate::{
    external_recipe::{
        error::ExternalRecipeError, ExternalRecipe, ExternalRecipeGetterTrait, ExternalRecipeStep,
        UrlMatch,
    },
    scraper::{Dom, ParentNode},
};

static PATH_REGEX: OnceLock<Regex> = OnceLock::new();

pub struct ExternalRecipeGetter;

#[async_trait]
impl ExternalRecipeGetterTrait for ExternalRecipeGetter {
    async fn get(&self, url: Url) -> Result<ExternalRecipe, ExternalRecipeError> {
        let response = reqwest::get(url).await?;
        let text = response.text().await?;
        let dom = Arc::new(Dom::create(text).await?);
        let mut steps = vec![];
        for step_element in dom.select_all(".recipe").await? {
            let mut ingredients = vec![];
            for ingredient_element in dom
                .select_all(".flex.items-start.justify-start.mb-1.space-x-3.text-lg.sm\\:text-base")
                .await?
            {
                ingredients.push(String::from(
                    ingredient_element.text_content().await?.trim(),
                ));
            }
            let images_wrap_element = dom.select(".images-wrap").await?;
            let mut files = vec![];
            for img_element in images_wrap_element.select_all("img").await? {
                files.push(format!(
                    "https://sallys-blog.de{}",
                    img_element.get_attribute("src").await?
                ));
            }
            steps.push(ExternalRecipeStep {
                ingredients,
                description: step_element.text_content().await?,
                files,
            });
        }
        Ok(ExternalRecipe {
            name: dom.select("h1").await?.text_content().await?,
            steps,
        })
    }

    fn url_matches(&self) -> Vec<UrlMatch<'static>> {
        vec![UrlMatch {
            schemes: &["http", "https"],
            domains: &["sallys-blog.de"],
            path_regex: PATH_REGEX.get_or_init(|| Regex::new(r"^/rezepte/.*$").unwrap()),
        }]
    }
}

File: src/services/external-recipe/modules/sallys-welt.ts
import type { ExternalRecipe } from "../../external-recipe.ts";
import {
  readContent,
  selectInParentNode,
  selectMultipleInParentNode,
} from "../getter.ts";

function get(data: string): ExternalRecipe {
  const parser = new DOMParser();
  const recipeDocument = parser.parseFromString(data, "text/html");

  return {
    name: readContent(selectInParentNode(recipeDocument, "h1")),
    steps: selectMultipleInParentNode(recipeDocument, ".recipe").map(
      (stepElement) => ({
        description: readContent(stepElement),
        ingredients: selectMultipleInParentNode(
          recipeDocument,
          ".flex.items-start.justify-start.mb-1.space-x-3.text-lg.sm\\:text-base",
        ).map((ingredientElement) => readContent(ingredientElement).trim()),
        files: [],
      }),
    ),
  };
}

/**
 * @type {ExternalRecipeModule}
 */
export { get };
